File: 2022/02.py
# https://adventofcode.com/2022/day/2

with open("02data.txt", "r") as datafile:
    input = datafile.readlines()

data = []
for line in input:
    data.append(line.split())

# Part 1

choicescore = {
    'X': 1, # rock
    'Y': 2, # paper
    'Z': 3, # scissors
}

def outcomescore(player, opponent):
    if (player == 'X' and opponent == 'A') \
    or (player == 'Y' and opponent == 'B') \
    or (player == 'Z' and opponent == 'C'):
        return 3 # draw
    elif (player == 'X' and opponent == 'C') \
      or (player == 'Y' and opponent == 'A') \
      or (player == 'Z' and opponent == 'B'):
        return 6 # player wins
    else:
        return 0 # player loses

score = 0
for round in data:
    score += choicescore.get(round[1])
    score += outcomescore(round[1], round[0])

print(score)

# Part 2

choicescore2 = {
    'A': 1, # rock
    'B': 2, # paper
    'C': 3, # scissors
}

pairwise = {
    'A': 'C', # rock beats scissors
    'B': 'A', # paper beats rock
    'C': 'B', # scissors beat paper
}

def fixround(outcome, opponent):
    if outcome == 'X': # lose
        return choicescore2[ pairwise[opponent] ]
    elif outcome == 'Y': # draw
        return choicescore2[opponent] + 3
    else: # 'Z' win - look up dict key by value
        for item in pairwise.items():
            if item[1] == opponent:
                return choicescore2[ item[0] ] + 6

score = 0
for round in data:
    score += fixround(round[1], round[0])

print(score)

File: 2022/02.dart
// https://adventofcode.com/2022/day/2
// Solved in Python then reimplemented in Dart to practise OOP

import 'dart:io';

enum Move {rock, paper, scissors}
enum Outcome {win, draw, lose}

class Round {
  late final Move _opponentMove, _playerMove;
  late final Outcome _outcome;
  late final int score;

  static const opponentMap = {
    'A': Move.rock,
    'B': Move.paper,
    'C': Move.scissors,
  };

  static const playerMap = {
    'X': Move.rock,
    'Y': Move.paper,
    'Z': Move.scissors,
  };

  static const outcomeMap = {
    'X': Outcome.lose,
    'Y': Outcome.draw,
    'Z': Outcome.win,
  };

  static const pairwise = {
    // key beats value
    Move.rock: Move.scissors,
    Move.paper: Move.rock,
    Move.scissors: Move.paper,
  };

  setOutcome() {
    if (_playerMove == _opponentMove) {
      _outcome = Outcome.draw;
    } else if (pairwise[_playerMove] == _opponentMove) {
      _outcome = Outcome.win;
    } else {
      _outcome = Outcome.lose;
    }
  }

  setPlayerMove() {
    if (_outcome == Outcome.lose) {
      _playerMove = pairwise[_opponentMove]!;
    } else if (_outcome == Outcome.draw) {
      _playerMove = _opponentMove;
    } else { // win
      _playerMove = pairwise.keys.firstWhere((elem) => pairwise[elem] == _opponentMove);
    }
  }

  int moveScore() {
    if (_playerMove == Move.rock) {
      return 1;
    } else if (_playerMove == Move.paper) {
      return 2;
    } else { // scissors
      return 3;
    }
  }

  int outcomeScore() {
    if (_outcome == Outcome.win) {
      return 6;
    } else if (_outcome == Outcome.draw) {
      return 3;
    } else { // lose
      return 0;
    }
  }

  Round(String opponentMove, {String? playerMove, String? outcome}) {
    _opponentMove = opponentMap[opponentMove]!;
    if (outcome == null) {
      _playerMove = playerMap[playerMove]!;
      setOutcome();
    }
    if (playerMove == null) {
      _outcome = outcomeMap[outcome]!;
      setPlayerMove();
    }
    score = moveScore() + outcomeScore();
  }
}

void main() {
  final List<String> input = File('02data.txt').readAsLinesSync();

  // Part 1
  int score = 0;
  input.forEach((line) {
    Round round = Round(line[0], playerMove: line[2]);
    score += round.score;
  });
  print(score);

  // Part 2
  score = 0;
  input.forEach((line) {
    Round round = Round(line[0], outcome: line[2]);
    score += round.score;
  });
  print(score);
}
